File: app/desktop/src/features/canonical/readModel/messageVisibility.ts
import type { CanonicalSessionMessage } from '@/kordi-app/types';
import { isExplicitPlaceholderSessionTitle } from '@/features/chat/sessionTitlePolicy';
import { parseCloudDirectMessageEnvelope } from '@/features/cloud/cloudDirectMessages';
import { isCloudAgentControlMessage } from '@/features/cloud/cloudAgentMessages';

function contentRecord(value: unknown): Record<string, unknown> {
  return value && typeof value === 'object' && !Array.isArray(value)
    ? value as Record<string, unknown>
    : {};
}

export function isPlaceholderSessionTitleNotice(message: CanonicalSessionMessage) {
  if (message.messageKind !== 'status') return false;
  const content = contentRecord(message.content);
  return content.kind === 'session-title-update'
    && content.scope === 'session'
    && typeof content.title === 'string'
    && isExplicitPlaceholderSessionTitle(content.title);
}

export function isSynchronizationOnlyCloudGroupTitleNotice(message: CanonicalSessionMessage) {
  if (message.messageKind !== 'status' || message.sourceTransport !== 'cloud-group-title-update') return false;
  const content = contentRecord(message.content);
  return content.kind === 'group-title-update'
    && content.scope === 'group'
    && content.synchronizationOnly === true
    && (content.sourceControlKind === 'group-invite' || content.sourceControlKind === 'group-update');
}

export function isInternalCloudAgentControlMessage(message: CanonicalSessionMessage) {
  const text = message.contentText.trim();
  const content = contentRecord(message.content);
  const normalized = content.schemaVersion === 1 && content.kind === 'message';
  const envelope = normalized ? null : parseCloudDirectMessageEnvelope(text);
  return content.synchronizationOnly === true
    || envelope?.synchronizationOnly === true
    || (Boolean(envelope) && message.messageKind === 'agent-model-change')
    || (!normalized && isCloudAgentControlMessage(text));
}

export function canonicalMessageCountsAsReadable(message: CanonicalSessionMessage) {
  if (message.sourceTransport === 'canonical-fork-snapshot') return false;
  if (['sending', 'processing'].includes(message.status.trim().toLowerCase())) return false;
  return !isPlaceholderSessionTitleNotice(message)
    && !isSynchronizationOnlyCloudGroupTitleNotice(message)
    && !isInternalCloudAgentControlMessage(message);
}

File: app/desktop/src-tauri/src/canonical_sessions/message_visibility.rs
//! Canonical read-cursor filtering for non-conversational message rows.

use rusqlite::{params, Connection, OptionalExtension};

pub(crate) fn latest_readable_session_message_id(
    conn: &Connection,
    session_id: &str,
) -> Result<Option<String>, String> {
    conn.query_row(
        "SELECT id
         FROM session_messages
         WHERE session_id = ?1
           AND COALESCE(source_transport, '') NOT IN ('canonical-fork-snapshot', 'cloud-group-fork-snapshot')
           AND LOWER(TRIM(status)) NOT IN ('sending', 'processing')
           AND NOT CASE
               WHEN LOWER(TRIM(COALESCE(message_kind, ''))) = 'status'
                    AND json_valid(content_json)
               THEN (
                   LOWER(TRIM(COALESCE(json_extract(content_json, '$.kind'), ''))) = 'session-title-update'
                   AND LOWER(TRIM(COALESCE(json_extract(content_json, '$.scope'), ''))) = 'session'
                   AND LOWER(TRIM(
                       CASE
                           WHEN SUBSTR(TRIM(COALESCE(json_extract(content_json, '$.title'), '')), 1, 1) = '#'
                               THEN SUBSTR(TRIM(COALESCE(json_extract(content_json, '$.title'), '')), 2)
                           ELSE COALESCE(json_extract(content_json, '$.title'), '')
                       END
                   )) IN ('new session', 'new chat', 'new fork', 'untitled session', 'session')
               ) OR (
                   LOWER(TRIM(COALESCE(source_transport, ''))) = 'cloud-group-title-update'
                   AND LOWER(TRIM(COALESCE(json_extract(content_json, '$.kind'), ''))) = 'group-title-update'
                   AND LOWER(TRIM(COALESCE(json_extract(content_json, '$.scope'), ''))) = 'group'
                   AND COALESCE(json_extract(content_json, '$.synchronizationOnly'), 0) = 1
                   AND LOWER(TRIM(COALESCE(json_extract(content_json, '$.sourceControlKind'), '')))
                       IN ('group-invite', 'group-update')
               )
               ELSE 0
           END
         ORDER BY sequence_num DESC, created_at_ms DESC
         LIMIT 1",
        params![session_id],
        |row| row.get(0),
    )
    .optional()
    .map_err(|err| err.to_string())
}
